scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -97,90 +105,135 @@ def highly_variable_genes(
     if min_disp is None: min_disp = 0.5
     if min_mean is None: min_mean = 0.0125
     if max_mean is None: max_mean = 3
+    if max_disp is None: max_disp = np.inf
 
-    X = np.expm1(adata.X) if flavor == 'seurat' else adata.X
-
-    mean, var = materialize_as_ndarray(_get_mean_var(X))
-    # now actually compute the dispersion
-    mean[mean == 0] = 1e-12  # set entries equal to zero to small value
-    dispersion = var / mean
-    if flavor == 'seurat':  # logarithmized mean as in Seurat
-        dispersion[dispersion == 0] = np.nan
-        dispersion = np.log(dispersion)
-        mean = np.log1p(mean)
-    # all of the following quantities are "per-gene" here
-    df = pd.DataFrame()
-    df['mean'] = mean
-    df['dispersion'] = dispersion
-    if flavor == 'seurat':
-        df['mean_bin'] = pd.cut(df['mean'], bins=n_bins)
-        disp_grouped = df.groupby('mean_bin')['dispersion']
-        disp_mean_bin = disp_grouped.mean()
-        disp_std_bin = disp_grouped.std(ddof=1)
-        # retrieve those genes that have nan std, these are the ones where
-        # only a single gene fell in the bin and implicitly set them to have
-        # a normalized disperion of 1
-        one_gene_per_bin = disp_std_bin.isnull()
-        gen_indices = np.where(one_gene_per_bin[df['mean_bin'].values])[0].tolist()
-        if len(gen_indices) > 0:
+    if batch_key is not None:
+        sanitize_anndata(adata)
+        batches = adata.obs[batch_key].cat.categories
+        df = []
+        for batch in batches:
```

Comment:
Would it be worth it to split calculation and argument handling into separate functions? I think it could clean up the flow control (and reduce indentation). Basically I'm thinking just put everything in the top level `else:` block into it's own function.